File: Exp 7 Application-based Algorithms using Greedy Approach /activity_selection.cpp
#include <iostream>
using namespace std;

const int MAX = 100;

void activitySelection(int start[], int finish[], int n) {
    // selection sort
    for (int i = 0; i < n - 1; i++) {
        int minPos = i;
        for (int j = i + 1; j < n; j++) {
            if (finish[j] < finish[minPos]) {
                minPos = j;
            }
        }
        // swap finish
        int tf = finish[i]; finish[i] = finish[minPos]; finish[minPos] = tf;
        // swap start
        int ts = start[i]; start[i] = start[minPos]; start[minPos] = ts;
    }

    cout << "\nSelected activities: ";
    int lastFinish = finish[0];
    cout << "1 ";  // first activity

    for (int i = 1; i < n; i++) {
        if (start[i] >= lastFinish) {
            cout << (i + 1) << " ";
            lastFinish = finish[i];
        }
    }
    cout << endl;
}

int main() {
    int n;
    cout << "Enter number of activities: ";
    cin >> n;

    int start[MAX], finish[MAX];
    cout << "Enter start and finish times:\n";
    for (int i = 0; i < n; i++) {
        cout << "Activity " << (i + 1) << ": ";
        cin >> start[i] >> finish[i];
    }

    activitySelection(start, finish, n);

    return 0;
}

File: Exp 7 Application-based Algorithms using Greedy Approach /activity_selection.py
MAX = 100


def activity_selection(start, finish, n):
    """Activity Selection using greedy approach"""
    # Selection sort
    for i in range(n - 1):
        min_pos = i
        for j in range(i + 1, n):
            if finish[j] < finish[min_pos]:
                min_pos = j
        # Swap finish
        finish[i], finish[min_pos] = finish[min_pos], finish[i]
        # Swap start
        start[i], start[min_pos] = start[min_pos], start[i]
    
    print("\nSelected activities: ", end="")
    last_finish = finish[0]
    print("1 ", end="")  # first activity
    
    for i in range(1, n):
        if start[i] >= last_finish:
            print(f"{i + 1} ", end="")
            last_finish = finish[i]
    print()


def main():
    n = int(input("Enter number of activities: "))
    
    start = []
    finish = []
    
    print("Enter start and finish times:")
    for i in range(n):
        times = input(f"Activity {i + 1}: ").split()
        start.append(int(times[0]))
        finish.append(int(times[1]))
    
    activity_selection(start, finish, n)


if __name__ == "__main__":
    main()
